test/test_torch.py

```diff
@@ -4385,37 +4385,40 @@ def test_narrow_empty(self, device):
             sz[d] = 0
             self.assertEqual(sz, y.size())
 
-    def test_index_copy(self, device):
-        num_copy, num_dest = 3, 20
-        dest = torch.randn(num_dest, 4, 5, device=device)
-        src = torch.randn(num_copy, 4, 5, device=device)
-        idx = torch.randperm(num_dest, device=device).narrow(0, 0, num_copy)
-        dest2 = dest.clone()
-        dest.index_copy_(0, idx, src)
-        for i in range(idx.size(0)):
-            dest2[idx[i]] = src[i]
-        self.assertEqual(dest, dest2, atol=0, rtol=0)
-
-        dest = torch.randn(num_dest, device=device)
-        src = torch.randn(num_copy, device=device)
-        idx = torch.randperm(num_dest, device=device).narrow(0, 0, num_copy)
-        dest2 = dest.clone()
-        dest.index_copy_(0, idx, src)
-        for i in range(idx.size(0)):
-            dest2[idx[i]] = src[i]
-        self.assertEqual(dest, dest2, atol=0, rtol=0)
-
-        # Bool tensor
-        dest = torch.zeros(2, 2, dtype=torch.bool, device=device)
-        src = torch.tensor([[True, True], [True, True]], device=device)
-        index = torch.tensor([0, 1], device=device)
-        dest.index_copy_(0, index, src)
-        self.assertEqual(dest, torch.tensor([[True, True], [True, True]], device=device))
-
+    @dtypes(*torch.testing.get_all_dtypes())
+    def test_index_copy(self, device, dtype):
+        num_copy, num_dest = 3, 5
+
+        def make_arg(batch_sizes, n, dim, contig):
+            size_arg = batch_sizes[:dim] + (n,) + batch_sizes[dim:]
+            arg = make_tensor(size_arg, device, dtype, low=None, high=None)
+            if contig:
+                arg = torch.testing.make_non_contiguous(arg)
+            return arg
+
+        # More thorough testing as in index_add
+        for dest_contig, src_contig, index_contig in product([True, False], repeat=3):
+            for other_sizes in ((), (4, 5)):
+                for dim in range(len(other_sizes)):
+                    dest = make_arg(other_sizes, num_dest, dim, dest_contig)
+                    src = make_arg(other_sizes, num_copy, dim, dest_contig)
+                    idx = torch.randperm(num_dest, dtype=torch.int64, device=device).narrow(0, 0, num_copy)
+                    if not index_contig:
+                        idx = torch.testing.make_non_contiguous(idx)
+                    dest2 = dest.clone()
+                    dest.index_copy_(dim, idx, src)
+                    for i in range(idx.size(0)):
+                        idx_dest = dim * (slice(None),) + (idx[i],)
+                        idx_src = dim * (slice(None),) + (i,)
+                        dest2[idx_dest] = src[idx_src]
+                    self.assertEqual(dest, dest2)
+
+    def test_errors_index_copy(self, device):
```

Comment:
The update to this test is good but it should validate the 3 cases in the kernels (2 on CPU, 1 on CUDA) that will throw index out of bound errors, too